### PR TITLE
Hotfix/update upload

### LINKED ISSUE #62: missing check on update upload whether the file name is already found in the folder
Currently we create an upload model with its ID and file name, and later updating it with the bucket and key of the file that is being uploaded.

There’s a bug where the file being uploaded has the same name as another file under the same folder it’s being uploaded to.

The bug is that there’s no check on the created upload whether the name of the file being uploaded is already found in that folder, so a user can only get an error on that at the end of the upload (potentially 5TB upload) when we complete it and use CreateFile which does check that and returns the error.

### PR DESCRIPTION
**Pull request by Shahar-Y**
fixes #62.
Added tests accordingly.
@omrishtam and @omerzamir should check their changes regarding the proto.

### CODE REVIEW

File: src/file/file.service.ts [hunk 1]
```diff
@@ -36,9 +36,18 @@ export class FileService {
    * @param bucket - is the s3 bucket in the storage
    * @param name - of the file uploaded
    */
-  public static async createUpload(key: string, bucket: string, name: string)
-  : Promise<IUpload> {
-    const upload: Partial<IUpload> = { key, bucket, name };
+  public static async createUpload(
```
Comment (omerzamir) [inline src/file/file.service.ts]: update the comment of the function - missing params

File: src/file/upload.repository.ts [hunk 2]
```diff
@@ -12,6 +13,8 @@ export class UploadRepository {
    * @param upload - the upload to be created.
    */
   static create(upload: Partial<IUpload>): Promise<IUpload> {
+    upload.ownerID = upload.ownerID ? upload.ownerID : 'null';
```
Comment (omrishtam) [inline src/file/upload.repository.ts]: Throw error if ownerID doesn't exist

File: src/file/upload.repository.ts [hunk 3]
```diff
@@ -1,5 +1,6 @@
 import { IUpload } from './upload.interface';
 import { uploadModel } from './upload.model';
+import { ObjectID } from 'mongodb';
```
Comment (omerzamir) [inline src/file/upload.repository.ts]: change to mongoose

File: src/file/upload.repository.ts (hunk 2 above)
Comment (omerzamir) [inline src/file/upload.repository.ts]: 'null' - Bug

File: src/file/upload.repository.ts [hunk 4]
```diff
@@ -36,8 +39,15 @@ export class UploadRepository {
    * @param key - the key of the upload. is unique with bucket.
    * @param bucket - the bucket of the upload.
    * @param uploadID - the new id.
+   * @param ownerID - owner of the file to be uploaded.
+   * @param parent - folder id of the file.
+   * @param filename - name of the file.
```
Comment (omerzamir) [inline src/file/upload.repository.ts]: extra params in comment, update it.

File: src/file/upload.repository.ts (hunk 3 above)
Comment (Shahar-Y) [reply src/file/upload.repository.ts]: mongoose doesn't seem to export this type. it uses mongodb's type in its core: 
image

Comment (omerzamir) [reply src/file/upload.repository.ts]: you used it already.
mongoose.Types.ObjectId()  ?

Comment (Shahar-Y) [reply src/file/upload.repository.ts]: this syntax doesn't work there

File: src/file/upload.repository.ts [hunk 5]
```diff
@@ -13,7 +13,9 @@ export class UploadRepository {
    * @param upload - the upload to be created.
    */
   static create(upload: Partial<IUpload>): Promise<IUpload> {
-    upload.ownerID = upload.ownerID ? upload.ownerID : 'null';
+    if (!upload.ownerID) {
```
Comment (omrishtam) [inline src/file/upload.repository.ts]: Should be `if (upload && !upload.ownerID) {...}`, since `upload` could be `undefined` or `null`.

File: src/file/file.rpc.ts [hunk 6]
```diff
@@ -52,15 +52,15 @@ export class RPC {
 
   // Generates a random key for the upload.
   private generateKey(call: any, callback: any) {
-    const key: string = FileService.generateKey();
+    const key:        string = FileService.generateKey();
```
Comment (omrishtam) [inline src/file/file.rpc.ts]: This file has an unused import of `IUser` in [line 3](https://github.com/meateam/file-service/blob/a1f222c6d57d630ee113eb08d6e1261698666e7c/src/file/file.rpc.ts#L3).

File: src/file/file.interface.ts [hunk 7]
```diff
@@ -11,7 +11,6 @@ export interface IFile {
   type: string;
   description?: string;
   ownerID: string;
-  owner?: IUser;
```
Comment (omrishtam) [inline src/file/file.interface.ts]: Remove unused import of `IUser` in [line 1](https://github.com/meateam/file-service/blob/a1f222c6d57d630ee113eb08d6e1261698666e7c/src/file/file.interface.ts#L1).

File: src/file/upload.repository.ts (hunk 3 above)
Comment (omrishtam) [reply src/file/upload.repository.ts]: This is the correct way to import and use `ObjectID`.

File: proto/file.proto [hunk 8]
```diff
@@ -20,6 +20,8 @@ service FileService {
 message CreateUploadRequest {
     string bucket = 1;
     string name = 2;
+    string ownerID = 4;
+    string parent = 5;
```
Comment (omrishtam) [inline proto/file.proto]: There are messages that have an `appID` field which is not defined in any of the schemas and is not used anywhere in the file-service's code. Remove it and compile the protobuf again.

Comment (omrishtam) [reply proto/file.proto]: Also remove `Bucket` field from `GenerateKeyRequest`

File: src/file/upload.repository.ts [hunk 9]
```diff
@@ -37,7 +42,11 @@ export class UploadRepository {
    * @param bucket - the bucket of the upload.
    * @param uploadID - the new id.
    */
-  static updateByKey(key: string, bucket: string, uploadID: string) {
+  static updateByKey(key: string, bucket: string, uploadID: string) : Promise<IUpload> {
     return uploadModel.findOneAndUpdate({ key, bucket }, { uploadID }, { new: true }).exec();
   }
+
+  static getMany(filter: Partial<IUpload>): Promise<IUpload[]> {
```
Comment (omerzamir) [inline src/file/upload.repository.ts]: Missing function documentation. (JSDocs)

File: src/file/upload.repository.ts (hunk 5 above)
Comment (omerzamir) [reply src/file/upload.repository.ts]: +1

File: src/file/file.rpc.ts [hunk 10]
```diff
@@ -52,29 +51,33 @@ export class RPC {
...
     callback(null, { key });
   }
 
   // Creates an upload object, present while uploading a file.
   private async createUpload(call: any, callback: any) {
-    const key: string = FileService.generateKey();
-    const bucket: string = call.request.bucket;
-    const name: string = call.request.name;
+    const key:        string = FileService.generateKey();
```
Comment (AllKinds) [inline src/file/file.rpc.ts]: Next time do style changes in a dedicated commits and branches.

File: src/file/upload.repository.ts (hunk 5 above)
Comment (AllKinds) [reply src/file/upload.repository.ts]: Then there should also be a check for the case when upload is `undefined` or `null`.
If that is the case there should be a Server Error.
